tss-esapi/src/structures/tagged/schemes.rs

```diff
@@ -61,3 +69,438 @@ impl TryFrom<TPMT_KEYEDHASH_SCHEME> for KeyedHashScheme {
+                    kdf2: hash_scheme.into(),
+                },
+            },
+            KeyDerivationFunctionScheme::Kdf1Sp800_108(hash_scheme) => TPMT_KDF_SCHEME {
+                scheme: KeyDerivationFunction::Kdf1Sp800_108.into(),
+                details: TPMU_KDF_SCHEME {
+                    kdf1_sp800_108: hash_scheme.into(),
+                },
+            },
+            KeyDerivationFunctionScheme::Mgf1(hash_scheme) => TPMT_KDF_SCHEME {
+                scheme: KeyDerivationFunction::Mgf1.into(),
+                details: TPMU_KDF_SCHEME {
+                    mgf1: hash_scheme.into(),
+                },
+            },
+            KeyDerivationFunctionScheme::Null => TPMT_KDF_SCHEME {
+                scheme: KeyDerivationFunction::Null.into(),
+                details: Default::default(),
+            },
+        }
+    }
+}
+
+impl TryFrom<TPMT_KDF_SCHEME> for KeyDerivationFunctionScheme {
+    type Error = Error;
+
+    fn try_from(tpmt_kdf_scheme: TPMT_KDF_SCHEME) -> Result<Self> {
+        match KeyDerivationFunction::try_from(tpmt_kdf_scheme.scheme)? {
+            KeyDerivationFunction::Kdf1Sp800_56a => Ok(KeyDerivationFunctionScheme::Kdf1Sp800_56a(
+                unsafe { tpmt_kdf_scheme.details.kdf1_sp800_56a }.try_into()?,
+            )),
+            KeyDerivationFunction::Kdf2 => Ok(KeyDerivationFunctionScheme::Kdf2(
+                unsafe { tpmt_kdf_scheme.details.kdf2 }.try_into()?,
+            )),
+            KeyDerivationFunction::Kdf1Sp800_108 => Ok(KeyDerivationFunctionScheme::Kdf1Sp800_108(
+                unsafe { tpmt_kdf_scheme.details.kdf1_sp800_108 }.try_into()?,
+            )),
+            KeyDerivationFunction::Mgf1 => Ok(KeyDerivationFunctionScheme::Mgf1(
+                unsafe { tpmt_kdf_scheme.details.mgf1 }.try_into()?,
+            )),
+            KeyDerivationFunction::Null => Ok(KeyDerivationFunctionScheme::Null),
+        }
+    }
+}
+
+/// Enum representing the rsa decryption scheme
+///
+/// # Details
+/// This corresponds to TPMT_RSA_DECRYPT.
+#[derive(Clone, Copy, Debug)]
+pub enum RsaDecryptionScheme {
+    RsaEs,
+    Oaep(HashScheme),
+    Null,
+}
+
+impl RsaDecryptionScheme {
+    /// Creates a new rsa decrypt scheme
+    pub fn create(
+        rsa_decrypt_algorithm: RsaDecryptAlgorithm,
+        hashing_algorithm: Option<HashingAlgorithm>,
+    ) -> Result<RsaDecryptionScheme> {
+        match rsa_decrypt_algorithm {
+            RsaDecryptAlgorithm::RsaEs => Ok(RsaDecryptionScheme::RsaEs),
+            RsaDecryptAlgorithm::Oaep => Ok(RsaDecryptionScheme::Oaep(HashScheme::new(
+                hashing_algorithm.ok_or_else(|| {
+                    error!("Hashing algorithm is required when creating RSA decrypt scheme of type OEAP");
+                    Error::local_error(WrapperErrorKind::ParamsMissing)
+                })?,
+            ))),
+            RsaDecryptAlgorithm::Null => Ok(RsaDecryptionScheme::Null),
+        }
+    }
```

Comment:
Actually, I just realised, because these are enums there's no easy way to block the variants and their content from being public, so `create` isn't that efficient (because you have to check for conditions that are more easily enforced when an enum variant is directly created - for example, for `RsaDecryptAlgorithm::Null` you don't check that `hashing_algorithm` is `None`), but maybe a `validate` method would be good if there's anything to validate. For example, in this case it seems like any hash works with `Oaep`, but otherwise `validate` would've been a good way to implement the checks.

Reply:
No, it is not possible to block 'bad' combinations if one creates the variants manually. So ```create``` is a convenience function that validates the combination. Yeah, it might be a good idea to split them up in two functions so it would be possible to validate a manually created variant.

Reply:
Ok, that's fair enough - I'm wondering how strict we want to be with the parameters. If we get two parameters that are inconsistent (because one isn't needed), do we just reject the call, or do we pretend the one that shouldn't have been there doesn't exist?

Reply:
Personally I want it to be really really hard to create something that seem to work but doesn't. So I like to be able to catch errors as soon as possible and be really strict to avoid errors later which only will make the TPM come back with one of its useless errors like 'wrong size associated with parameter 2'.

Reply:
I have updated so that it gives an error for if a hashing algorithm have been provided when none is expected in a lot of places.